File: 백준/Gold/2342. Dance Dance Revolution/Dance Dance Revolution.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    private static final int INF = (int) 1e9;
    static int[][][] dp;
    static int cmd, res;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        dp = new int[st.countTokens()][5][5];

        for (int[][] lefts : dp) {
            for (int[] right : lefts) {
                Arrays.fill(right, INF);
            }
        }

        dp[0][0][0] = 0;
        res = 0;

        int i = 0;

        while (++i < dp.length) {
            cmd = Integer.parseInt(st.nextToken());

            if (cmd == 0) break;

            for (int lf = 0; lf <= 4; lf++) {
                for (int rf = 0; rf <= 4; rf++) {
                    if (dp[i - 1][lf][rf] == INF) continue;

                    dp[i][lf][cmd] = Math.min(dp[i][lf][cmd], dp[i - 1][lf][rf] + dist(rf, cmd));
                    dp[i][cmd][rf] = Math.min(dp[i][cmd][rf], dp[i - 1][lf][rf] + dist(lf, cmd));
                }
            }
        }

        int res = INF;

        for (int[] row : dp[i - 1]) {
            for (int v : row) {
                res = Math.min(res, v);
            }
        }

        System.out.println(res);
        br.close();
    }

    static int dist(int u, int v) {
        if (u == v) return 1;

        if (u == 0) return 2;

        if (Math.abs(u - v) == 2) return 4;

        return 3;
    }
}

File: 백준/Gold/2342. Dance Dance Revolution/Dance Dance Revolution.kt
import java.io.StreamTokenizer
import kotlin.math.abs

private val st = StreamTokenizer(System.`in`.bufferedReader())

private fun readInt(): Int {
    st.nextToken()
    return st.nval.toInt()
}

fun main() {
    val ddr = DDR()
    var next = readInt()

    while (next > 0) {
        ddr.move(next)

        next = readInt()
    }

    with(System.out.bufferedWriter()) {
        append("${ddr.cost}")
        flush()
        close()
    }
}

class DDR {
    val INF = 1e9.toInt()
    val dp = Array(100_001) { Array(5) { IntArray(5) { INF } } }
    var count = 0
    val cost
        get() = dp[count].minOf { it.min() }

    init {
        dp[0][0][0] = 0
    }

    fun move(next: Int) {
        count++

        for (i in 0..4) {
            for (j in 0..4) {
                if (i > 0 && i == j) continue

                val dist = dist(j, next)

                dp[count][i][next] = minOf(dp[count][i][next], dp[count - 1][i][j] + dist)
                dp[count][next][i] = minOf(dp[count][next][i], dp[count - 1][j][i] + dist)
            }
        }
    }

    fun dist(u: Int, v: Int): Int {
        return when {
            u == 0 -> 2
            u == v -> 1
            abs(u - v) == 2 -> 4
            else -> 3
        }
    }
}